src/Parser.js

```diff
@@ -600,80 +593,148 @@ Parser.prototype.parseArguments = function(func, funcData) {
-        this.pos = whitespace.position;
+        while (this.nextToken.text === " ") {
+            this.consume();
+        }
     }
-
     // By the time we get here, innerMode is one of "text" or "math".
     // We switch the mode of the parser, recurse, then restore the old mode.
-    this.mode = innerMode;
-    this.nextToken = this.lexer.lex(this.pos, innerMode);
     var res;
     if (optional) {
         res = this.parseOptionalGroup();
     } else {
         res = this.parseGroup();
     }
-    this.mode = outerMode;
-    this.nextToken = this.lexer.lex(this.pos, outerMode);
+    this.switchMode(outerMode);
     return res;
 };
 
 /**
+ * Parses a group where the string formed by the brace-enclosed tokens
+ * must conform to some regular expression.
+ *
+ * @param modeName {string} Used to describe the mode in error messages
+ * @param optional {boolean} Whether the group is optional or required
+ * @param reOnline {!RegExp} This regular expression must match
+ *     the final result as well as any non-empty prefix of a final result
+ * @param reFinal {!RegExp} This regular expression must match
+ *     the final result, and that match will be returned
+ */
+Parser.prototype.parseSpecialGroup = function(modeName, optional,
+                                              reOnline, reFinal) {
+    if (optional && this.nextToken.text !== "[") {
+        return null;
+    }
+    var outerMode = this.mode;
+    this.mode = "text";
+    this.expect(optional ? "[" : "{");
+    var str = "";
+    var firstToken = this.nextToken;
+    var lastToken = firstToken;
+    while (this.nextToken.text !== (optional ? "]" : "}")) {
+        lastToken = this.nextToken;
+        str += lastToken.text;
+        if (!reOnline.test(str)) {
+            throw new ParseError(
+                "Invalid " + modeName + ": '" + str + "…'",
+                firstToken.range(lastToken, str));
+        }
+        this.consume();
+    }
+    var match = reFinal.exec(str);
+    if (!match) {
+        throw new ParseError(
+            "Invalid " + modeName + ": '" + str + "'",
+            firstToken.range(lastToken, str));
+    }
+    this.mode = outerMode;
+    this.expect(optional ? "]" : "}");
+    return {
+        match: match,
+        range: firstToken.range(lastToken, str),
+    };
+};
+
+/**
+ * Parses a color description.
+ */
+Parser.prototype.parseColorGroup = function(optional) {
+    var res = this.parseSpecialGroup(
+        "color", optional,
+        /^(#[a-z0-9]*|[a-z]+)$/i,
+        /^(#[a-z0-9]+|[a-z]+)$/i);
+    if (!res) {
+        return null;
+    }
+    return new ParseFuncOrArgument(
+        new ParseNode("color", res.match[0], this.mode),
+        false);
+};
+
+/**
+ * Parses a size specification, consisting of magnitude and unit.
+ */
+Parser.prototype.parseSizeGroup = function(optional) {
+    var res = this.parseSpecialGroup(
+        "size", optional,
+        /^ *(-?) *(?:(?:\d+(?:\.\d*)?|\.\d+) *[a-z]{0,2} *)?$/,
```

Comment:
My goal was to have the error reporting in one place, and also to catch ill-formed special groups pretty early, thanks to that `reOnline` parameter. But I do confess that I see no compelling reason to do so. If one were to do this naively, we might get strange error messages: `\rule{2{cm}}` would complain abouut `2{cm` being an invalid length, instead of `2{…` being an invalid length. But I don't think this is a real problem, and wouldn't try to support nested groups unless there was some good reason to do so.